Add E2E test for tracking opt-in toggle

circlecube opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #274 

## Relevant technical choices

Adds tests to verify default opt-in value and tracking is in place as well as verify toggling opt-in setting removes tracking. 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

aaemnnosttv commented:
@googlebot I consent.

swissspidy commented on bin/local-env/install-wordpress.sh:
```diff
@@ -94,6 +94,10 @@ docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm $CONTAINER touch /var/www/h
 
+# Set pretty permalinks.
+echo -e $(status_message "Setting permalink structure...")
+docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm -u 33 $CLI rewrite structure '%postname%' --hard
```
Curious: why was this needed?

If the tests fail without pretty permalinks that sounds like a bug to me.

aaemnnosttv replied:
It's for the predicate function used in `page.waitForResponse()`, otherwise the REST url uses all query parameters which would require more parsing.

I agree that it should work either way, but we need to identify that request specifically and it both seemed easier and likely more common in the real world to use pretty permalinks.

swissspidy replied:
"More parsing" meaning using `reqURL.params.get()` instead of `reqURL.pathname`? 🤔 

But yeah I guess it's fine to keep this. It shouldn't bite us in the future.

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+	beforeEach( async() => {
+		await visitAdminPage( 'admin.php', 'page=googlesitekit-settings' );
+		await page.waitForSelector( '.mdc-tab-scroller__scroll-content' );
+		page.click( '.mdc-tab-scroller__scroll-content button:last-child' );
```
Without going through this manually, one has no idea what this clicks on.

Let's use the same approach as in #343 for this for consistency reasons:

```js
await visitAdminPage( 'admin.php', 'page=googlesitekit-settings' );
await page.waitForSelector( '.mdc-tab-scroller__scroll-content' );

// Click on Admin Settings Tab.
await expect( page ).toClick( 'button.mdc-tab', { text: 'Admin Settings' } );
```

circlecube replied:
Thanks for the consistency! This is updated.

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+describe( 'Providing client configuration', () => {
+
+	beforeAll( async() => {
+		await activatePlugin( 'google-site-kit' );
```
The Site Kit plugin should actually be already active during E2E tests (done in the Docker setup). The only exception where it needs to be (de)activated is in the test for the plugin activation itself.

circlecube replied:
That makes perfect sense, no longer activating/deactivating the plugin.

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+
+	afterAll( async() => {
+		await resetSiteKit();
+		await deactivatePlugin( 'google-site-kit' );
```
As mentioned above, no need to deactivate the plugin.

circlecube replied:
Fixed here too

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+		await deactivatePlugin( 'e2e-tests-auth-plugin' );
+	} );
+
+	it( 'Should select opt-in by default', async() => {
```
We currently begin the test names with lowercase letters, let's do this here as well for consistency reasons.

circlecube replied:
All test names update to lower case.

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+
+		await page.waitForSelector( '#opt-in' );
+
+		expect( await page.$eval( '#opt-in', el => el.checked ) ).toBe( true );
```
We currently successfully use `toMatchElement` to make these assertions simpler to read and write. Example here:

```js
await expect( page ).toMatchElement( '#opt-in[checked]' );
```

aaemnnosttv replied:
That matches the html of the element when it was rendered on page load, not the current state.

If the toggle was enabled when the page loaded, it will have a `checked` attribute, but otherwise it will not. This attribute does not change when the input is toggled, that's why we're checking the property on the element.

swissspidy replied:
Ah, yes of course. Thanks for clarification :-)

circlecube replied:
Updated a couple that could be simplified when the checkbox did not toggle.

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+		expect( await page.$eval( '#opt-in', el => el.checked ) ).toBe( true );
+
+	} );
+	it( 'Should have tracking code when opted in', async() => {
```
Here there is no empty line between the tests

swissspidy commented on tests/e2e/specs/admin-tracking.test.js:
```diff
@@ -0,0 +1,95 @@
+
+		await page.$eval( '#opt-in', elem => elem.click() );
+
+		await page.waitForResponse( res => {
```
Nice choice here!

circlecube commented:
Thanks for the thoughtful review @swissspidy - I've addressed all feedback I believe and we're just waiting for your response regarding the permalinks above.
